[c10d] Fix the device query story of ProcessGroup

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136790

Function `_get_pg_default_device` is being used outside of `distributed_c10d.py`.

A concern is that people may not be aware of what it actually does, due to bad naming of this function:
`Return the device to use with ``group`` for control flow usage (object collectives, barrier).`

The remediation is as follows:
- Added a deprecation warning to `_get_pg_default_device`;
- Added a private function `_get_object_coll_device` to undertake what it does;
- Added a `_device_capability` function for users who want to query the device support of a PG -- it returns a plain list, no more "default" choice.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC 

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[c10d] Fix the device query story of ProcessGroup

[ghstack-poisoned]

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -750,9 +745,10 @@ def _check_valid_timeout(timeout: Any) -> None:
+def _get_object_coll_device(group: Optional[ProcessGroup] = None) -> str:
     """
-    Return the device to use with ``group`` for control flow usage (object collectives, barrier).
+    Return the device type to use with ``group`` for control flow usage (object
```
i'm still not sure i understand what the point of this API is, or how its supposed to be used.  ok let me read the callsites first and come back..

wconstab replied:
ok- iiuc, the issue is that object collective APIs are unique in accepting non-tensor objects which do not have a 'device' attribute to key off of.  Absent that device argument, we look to the PG to determine which device to use.  So having a helper to determine a backend kinda makes sense.  

(or, we could have just added a 'device' argument to obj coll APIs, and made it an error if the user did not specify device and the PG had multiple backends!)

I don't know what "for control flow usage" means either??

kwen2501 replied:
Yep, that's right. 
Removed the "control flow usage" wording from the description.

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -766,13 +762,69 @@ def _get_pg_default_device(group: Optional[ProcessGroup] = None) -> torch.device
+
+def _get_pg_default_device(group: Optional[ProcessGroup] = None) -> torch.device:
+    """
+    .. note:: This method may be deprecated, it only stays for
```
may be deprecated? haha

kwen2501 replied:
will be deprecated :)

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -766,13 +762,69 @@ def _get_pg_default_device(group: Optional[ProcessGroup] = None) -> torch.device
+    .. note:: This method may be deprecated, it only stays for
+        backward-compatiblity reason. Alternatives:
+
+        - If you need to find a device for object collecitves, please use
```
typo collectives

kwen2501 replied:
Fixed.

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -750,9 +745,10 @@ def _check_valid_timeout(timeout: Any) -> None:
 
 
-def _get_pg_default_device(group: Optional[ProcessGroup] = None) -> torch.device:
+def _get_object_coll_device(group: Optional[ProcessGroup] = None) -> str:
```
can you say more about what the users are doing with this API outside of this file?  

1) renaming it won't make people stop using the new name
2) its an `_` api anyway, so we are free to change it i guess?

kwen2501 replied:
Here is a search of references:
https://github.com/search?q=repo%3Apytorch%2Fpytorch%20_get_pg_default_device&type=code
Mainly in DCP.

kwen2501 replied:
Yeah, renaming or `_` won't stop people from using it. I just want to make the name clear so users know what it does.

wconstab replied:
tbh i feel like 'default device' was more clear and i dont really know what 'object_coll_device' means at all.  (unless i read the code for object coll, realize that it needs to infer a device without a tensor to look at).

Do you think it makes sense to just deprecate the fn without renaming it?  idk..

kwen2501 replied:
Sorry, I am a bit confused. I am deprecating `_get_pg_default_device`. But I also need a function to hold the logic for inferring a device for object coll. 

The `dont really know what 'object_coll_device' means at all` seems like a bonus because then people won't easily use it outside.

H-Huang replied:
> Here is a search of references:
https://github.com/search?q=repo%3Apytorch%2Fpytorch%20_get_pg_default_device&type=code
Mainly in DCP.

It looks like DCP and FSDP and mostly using similar to the `_device_capability()` method you added. So maybe that should be updated?

kwen2501 replied:
Yeah, thanks for mentioning that. I was in side chat with @LucasLLC @wz337 @fegin earlier. Maybe we should file an issue to migrate the usage?

kwen2501 committed:
Update on "[c10d] Fix the device query story of ProcessGroup"


Function `_get_pg_default_device` is being used outside of `distributed_c10d.py`.

A concern is that people may not be aware of what it actually does, due to bad naming of this function:
`Return the device to use with ``group`` for control flow usage (object collectives, barrier).`

The remediation is as follows:
- Added a deprecation warning to `_get_pg_default_device`;
- Added a private function `_get_object_coll_device` to undertake what it does;
- Added a `_device_capability` function for users who want to query the device support of a PG -- it returns a plain list, no more "default" choice.

cc H-Huang wz337 LucasLLC fegin 

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

zeshengzong commented on torch/distributed/distributed_c10d.py:
```diff
@@ -798,25 +794,113 @@ def _get_pg_default_device(group: Optional[ProcessGroup] = None) -> torch.device
+    if len(devices) == 1:
+        # User fixed exactly one backend in `init_process_group`
+        return devices[0]
+    elif len(devices) == 0:
+        raise RuntimeError(
+            "Default device not found, because no backend has been registered "
+            "with this ProcessGroup."
+        )
```
Hello, I saw previous logic is return `torch.device("cpu")` in here, is it better to keep the same if this purely for BC ? :D

kwen2501 replied:
Yeah, my major purpose is still to deprecate this function, so whether to keep the semantics the same is not very important. The lines here says "I don't want to guess because I don't know." :)

kwen2501 committed:
Update on "[c10d] Fix the device query story of ProcessGroup"


Function `_get_pg_default_device` is being used outside of `distributed_c10d.py`.

A concern is that people may not be aware of what it actually does, due to bad naming of this function:
`Return the device to use with ``group`` for control flow usage (object collectives, barrier).`

The remediation is as follows:
- Added a deprecation warning to `_get_pg_default_device`;
- Added a private function `_get_object_coll_device` to undertake what it does;
- Added a `_device_capability` function for users who want to query the device support of a PG -- it returns a plain list, no more "default" choice.

cc H-Huang wz337 LucasLLC fegin 

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge -f "failure seems to be unrelated"
